### ISSUE TITLE
How to reproduce the result

### ISSUE THREAD

**Issue by ShaoShuai0605** [ZONG0004/MacroHFT#2]
Hello, When I want to reproduce the result, there might be some problem because the return on the validation and test is much worse than yours. Are there any tips or specific settings?

**Comment by ZONG0004** [comment]
I am very sorry for the late reply. I think there might be some issues in training sub-agents because of Nvidia and CUDA library updates.  We have uploaded our trained sub-agent checkpoints so that you can directly use them to train the high-level agent. We also remind you that by running the high-level training script, the output only shows results on the validation set. You can select the best epoch manually or use our saved model for high-level agent to run on the testing set. The inference function can be found in the high-level training script. 
I hope this can address your problems.

**Comment by Dreamlikec** [comment]
I also doubt whether anyone has truly replicated the results in the paper. The results I obtained have always been unsatisfactory.I train a long time and still got the profit margin negative. And I saw that you upload is slow-level checkpoint, why you say "directly use them to train high-level agent"? I can't understand.

**Comment by ZONG0004** [comment]
> I also doubt whether anyone has truly replicated the results in the paper. The results I obtained have always been unsatisfactory.I train a long time and still got the profit margin negative. And I saw that you upload is slow-level checkpoint, why you say "directly use them to train high-level agent"? I can't understand.

Thanks for your comment. I have also thoroughly investigated the discrepancy in sub-agent training. However, by using the same code, there are still some small differences during the training procedure, which might accumulate to a different trading behavior. I have no other choice but to conclude that this might be caused by Nvidia and CUDA library updates (maybe a CUDA 11 to 12). 
To remedy this problem, I uploaded the checkpoints of sub-agents, which can be directly loaded in the high-level training python script. (You can refer to line 129 to 150 in RL/agent/high_level.py) In this case, at least the high-level training can be reproduced. 
I am very sorry for the confusion caused to you,  but maybe you can try to run the high-level training script and check the outcomes, which should be the same as we claimed in the paper. It is also worth noticing that the return rates in the log file of high-level training are for training and validation sets. You can choose the epoch (which I believe should be epoch 15) with the highest validation return and conduct inference on testing set. 
Also feel free to come back to me if there is still any problem or if you have any concerns.

**Comment by Dreamlikec** [comment]
> > I also doubt whether anyone has truly replicated the results in the paper. The results I obtained have always been unsatisfactory.I train a long time and still got the profit margin negative. And I saw that you upload is slow-level checkpoint, why you say "directly use them to train high-level agent"? I can't understand.
> 
> Thanks for your comment. I have also thoroughly investigated the discrepancy in sub-agent training. However, by using the same code, there are still some small differences during the training procedure, which might accumulate to a different trading behavior. I have no other choice but to conclude that this might be caused by Nvidia and CUDA library updates (maybe a CUDA 11 to 12). To remedy this problem, I uploaded the checkpoints of sub-agents, which can be directly loaded in the high-level training python script. (You can refer to line 129 to 150 in RL/agent/high_level.py) In this case, at least the high-level training can be reproduced. I am very sorry for the confusion caused to you, but maybe you can try to run the high-level training script and check the outcomes, which should be the same as we claimed in the paper. It is also worth noticing that the return rates in the log file of high-level training are for training and validation sets. You can choose the epoch (which I believe should be epoch 15) with the highest validation return and conduct inference on testing set. Also feel free to come back to me if there is still any problem or if you have any concerns.

Thank you so much for your quick response! I think I understand the purpose of uploading the low-level checkpoint, but there are a few questions that may continue to trouble everyone:

1. If I want to train low-level data for a different time range or a different cryptocurrency (not your paper give), can this still be reproduced using your code?
2. Where did you obtain your data source? (If possible, could you recommend a source?)

**Comment by ZONG0004** [comment]
> > > I also doubt whether anyone has truly replicated the results in the paper. The results I obtained have always been unsatisfactory.I train a long time and still got the profit margin negative. And I saw that you upload is slow-level checkpoint, why you say "directly use them to train high-level agent"? I can't understand.
> > 
> > 
> > Thanks for your comment. I have also thoroughly investigated the discrepancy in sub-agent training. However, by using the same code, there are still some small differences during the training procedure, which might accumulate to a different trading behavior. I have no other choice but to conclude that this might be caused by Nvidia and CUDA library updates (maybe a CUDA 11 to 12). To remedy this problem, I uploaded the checkpoints of sub-agents, which can be directly loaded in the high-level training python script. (You can refer to line 129 to 150 in RL/agent/high_level.py) In this case, at least the high-level training can be reproduced. I am very sorry for the confusion caused to you, but maybe you can try to run the high-level training script and check the outcomes, which should be the same as we claimed in the paper. It is also worth noticing that the return rates in the log file of high-level training are for training and validation sets. You can choose the epoch (which I believe should be epoch 15) with the highest validation return and conduct inference on testing set. Also feel free to come back to me if there is still any problem or if you have any concerns.
> 
> Thank you so much for your quick response! I think I understand the purpose of uploading the low-level checkpoint, but there are a few questions that may continue to trouble everyone:
> 
> 1. If I want to train low-level data for a different time range or a different cryptocurrency (not your paper give), can this still be reproduced using your code?
> 2. Where did you obtain your data source? (If possible, could you recommend a source?)

For the first question, I would say it depends. Generally speaking, the key idea of low-level agent training is to provide a more systematic way of handling context features instead of just concatenating them to snapshot features. In this case, I think it can help RL training for quantitative trading, which could be called reproducible. Besides the paper mentioned testing period, we backtest the final model on ETHUSDT from 2023/11/01 to 2024/01/31, which gets 31.93% return rate over buy-and-hold which is 24.44%. However, change of time range, trading frequency and trading pair can also lead to negative profit, especially for cryptocurrency market from 2024 till now, which is much more violated. Based on my experience, RL-based trading algorithms should be treated carefully case by case, and adjustments might be necessary when the market condition changes a lot. 

For the second question, we bought the data from tardis, one of the overseas cryptocurrency data providers. I believe there are also others who provide this kind of service.
